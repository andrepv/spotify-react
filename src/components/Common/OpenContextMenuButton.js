import React, { Component } from "react";
import PropTypes from "prop-types";

export default class OpenContextMenuButton extends Component {
  constructor(props) {
    super(props);
    this.toggleContextMenu = this.toggleContextMenu.bind(this);
    this.closeContextMenu = this.closeContextMenu.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
  }

  state = {
    isContextMenuOpen: false,
  }

  componentWillUnmount() {
    this.detachEventListeners();
  }

  handleClickOutside(e) {
    if (!this.button.contains(e.target)) {
      this.closeContextMenu();
    }
  }

  toggleContextMenu() {
    this.setState({
      isContextMenuOpen: !this.state.isContextMenuOpen,
    });
    if (!this.state.isContextMenuOpen) {
      this.attachEventListeners();
    } else {
      this.detachEventListeners();
    }
  }

  closeContextMenu() {
    this.setState({
      isContextMenuOpen: false,
    });
    this.detachEventListeners();
  }

  attachEventListeners() {
    document.addEventListener("click", this.handleClickOutside);
    window.addEventListener("scroll", this.closeContextMenu);
  }

  detachEventListeners() {
    document.removeEventListener("click", this.handleClickOutside);
    window.removeEventListener("scroll", this.closeContextMenu);
  }

  render() {
    const {renderContent, renderContextMenu, className = ""} = this.props;
    const {isContextMenuOpen} = this.state;
    const btnClassName = `more-btn ${className} ${
      isContextMenuOpen ? "more-btn_active" : ""
    }`;
    return (
      <button
        className={btnClassName}
        ref={el => (this.button = el)}
      >
        {renderContent(this.toggleContextMenu)}
        {isContextMenuOpen && renderContextMenu(this.closeContextMenu)}
      </button>
    );
  }
}

OpenContextMenuButton.propTypes = {
  renderContextMenu: PropTypes.func.isRequired,
  renderContent: PropTypes.func.isRequired,
  className: PropTypes.string,
};