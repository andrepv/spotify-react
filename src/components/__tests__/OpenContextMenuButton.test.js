import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import OpenContextMenuButton from "components/Common/OpenContextMenuButton";

describe("<OpenContextMenuButton />", () => {
  it("renders correctly", () => {
    const props = {
      renderContextMenu: closeContextMenu => {
        return (
          <div onClick={closeContextMenu}>
            Menu
          </div>
        );
      },
      renderContent: toggleContextMenu => {
        return (
          <span onClick={toggleContextMenu}>
            Open Menu
          </span>
        );
      },
      className: "test",
    };
    const component = renderer.create(
      <OpenContextMenuButton {...props} />
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it("should toggle context menu correctly", () => {
    const props = {
      renderContextMenu: jest.fn(),
      renderContent: jest.fn(),
    };
    const wrapper = shallow(<OpenContextMenuButton {...props}/>);
    expect(props.renderContent.mock.calls.length).toBe(1);
    expect(props.renderContent).lastCalledWith(expect.any(Function));

    wrapper.setState({isContextMenuOpen: true});
    expect(props.renderContextMenu.mock.calls.length).toBe(1);
    expect(props.renderContextMenu).lastCalledWith(expect.any(Function));
    expect(wrapper.find(".more-btn_active")).toHaveLength(1);

    wrapper.setState({isContextMenuOpen: false});
    expect(wrapper.find(".more-btn_active")).toHaveLength(0);
    expect(props.renderContextMenu.mock.calls.length).toBe(1);
  });
});