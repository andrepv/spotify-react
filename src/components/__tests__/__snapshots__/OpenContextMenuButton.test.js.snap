// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OpenContextMenuButton /> renders correctly 1`] = `
<button
  className="more-btn test "
>
  <span
    onClick={[Function]}
  >
    Open Menu
  </span>
</button>
`;
